### PR TITLE
[RHBA-577] Create OpenShift templates based on RHBA LA release -- 1st draft

### PR DESCRIPTION
**Pull request by calvinzhuca**
Thanks for submitting your Pull Request!

https://issues.jboss.org/browse/RHBA-577

Please make sure your PR meets the following requirements:

- [x] Pull Request title is properly formatted: `[CLOUD-XYA] Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Attached commits represent units of work and are properly formatted
- [x] You have read and agreed to the Developer Certificate of Origin (DCO) (see `CONTRIBUTING.md`)
- [x] Every commit contains `Signed-off-by: Your Name <yourname@example.com>` - use `git commit -s`

### CODE REVIEW

File: templates/rhba70-businesscentral-monitoring-with-smartrouter.yaml [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+  name: KIE_SERVER_CONTROLLER_USER
+  value: controllerUser
+  required: false
+- displayName: KIE Server Controller Password
+  description: KIE server controller password (Sets the org.kie.server.controller.pwd system property)
+  name: KIE_SERVER_CONTROLLER_PWD
+  from: "[a-zA-Z]{6}[0-9]{1}!"
+  generate: expression
+  required: false
+- displayName: Smart router controller token
```
Comment (jakubschwan) [inline templates/rhba70-businesscentral-monitoring-with-smartrouter.yaml]: This display name do not look correct. Shouln't it be Kie Server Controller token?

File: templates/rhba70-businesscentral-monitoring.yaml [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+- kind: Service
+  apiVersion: v1
+  spec:
+    ports:
+    - port: 8080
+      targetPort: 8080
+    selector:
+      deploymentConfig: "${APPLICATION_NAME}-rhbacentrmon"
+  metadata:
+    name: "${APPLICATION_NAME}-rhbacentrmonmon"
```
Comment (jakubschwan) [inline templates/rhba70-businesscentral-monitoring.yaml]: At this line there is a typo "monmon"

File: templates/rhba70-businesscentral.yaml [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+  required: false
+- displayName: ImageStream Namespace
+  description: Namespace in which the ImageStreams for Red Hat Middleware images are
+    installed. These ImageStreams are normally installed in the openshift namespace.
+    You should only need to modify this if you've installed the ImageStreams in a
+    different namespace/project.
+  name: IMAGE_STREAM_NAMESPACE
+  value: openshift
+  required: true
+- displayName: MySQL Image Stream Tag
```
Comment (jakubschwan) [inline templates/rhba70-businesscentral.yaml]: This template is only for Business Central and there is no need for a DB so all environment variables for DB can be removed.

File: templates/rhba70-businesscentral.yaml [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+    selector:
+      deploymentConfig: "${APPLICATION_NAME}-rhbacentr"
+  metadata:
+    name: secure-${APPLICATION_NAME}-rhbacentr
+    labels:
+      application: "${APPLICATION_NAME}"
+    annotations:
+      description: The Business Central web server's https port.
+
+
```
Comment (jakubschwan) [inline templates/rhba70-businesscentral.yaml]: empty lines can be removed

File: templates/rhba70-full-mysql.yaml [hunk 5]
```diff
@@ -0,0 +1,693 @@
...
+            containerPort: 8778
+            protocol: TCP
+          - name: http
+            containerPort: 8080
+            protocol: TCP
+          - name: https
+            containerPort: 8443
+            protocol: TCP
+          env:
+          - name: KIE_ADMIN_PWD
```
Comment (ruromero) [inline templates/rhba70-full-mysql.yaml]: The KIE_ADMIN_PWD and KIE_ADMIN_USER should use the same values as the MAVEN_REPO_PASSWORD and MAVEN_REPO_USERNAME respectively or else they will have different generated values

File: templates/rhba70-businesscentral.yaml (hunk 3 above)
Comment (calvinzhuca) [reply templates/rhba70-businesscentral.yaml]: fixed by taking out the DB section for Business Central

File: templates/rhba70-full-mysql.yaml (hunk 5 above)
Comment (calvinzhuca) [reply templates/rhba70-full-mysql.yaml]: Fixed in rhba70-full-mysql.yaml and rhba70-full-mysql-persistent.yaml
          - name: MAVEN_REPO_USERNAME
            value: "${KIE_ADMIN_USER}"
          - name: MAVEN_REPO_PASSWORD
            value: "${KIE_ADMIN_PWD}"

File: templates/rhba70-businesscentral-monitoring-with-smartrouter.yaml (hunk 1 above)
Comment (calvinzhuca) [reply templates/rhba70-businesscentral-monitoring-with-smartrouter.yaml]: fixed the copy & paste error

File: templates/rhba70-businesscentral-monitoring.yaml (hunk 2 above)
Comment (calvinzhuca) [reply templates/rhba70-businesscentral-monitoring.yaml]: fixed

File: templates/rhba70-businesscentral.yaml (hunk 4 above)
Comment (calvinzhuca) [reply templates/rhba70-businesscentral.yaml]: removed empty lines

**Comment by calvinzhuca** [comment]
JIRA is at https://issues.jboss.org/browse/BAPL-691

File: templates/rhba70-full-mysql.yaml (hunk 5 above)
Comment (jakubschwan) [reply templates/rhba70-full-mysql.yaml]: This configuration is not valid for case, when is connected an external maven repository. For external maven repository can be set different user and password.

Comment (ruromero) [reply templates/rhba70-full-mysql.yaml]: @jakubschwan then the maven admin shouldn't have a default and the password not be generated. The original approach prevents the template from working. The description already provides enough information for the user.
@calvinzhuca given @jakubschwan's point you could leave the old values and remove the default value for the admin and password generation, leaving up to the template user to decide which values to provide

File: templates/rhba70-businesscentral.yaml [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+    selector:
+      deploymentConfig: "${APPLICATION_NAME}-rhbacentr"
+    template:
+      metadata:
+        name: "${APPLICATION_NAME}-rhbacentr"
+        labels:
+          deploymentConfig: "${APPLICATION_NAME}-rhbacentr"
+          application: "${APPLICATION_NAME}"
+      spec:
+        serviceAccountName: businesscentral-service-account
```
Comment (ruromero) [inline templates/rhba70-businesscentral.yaml]: The service account should be also created in the template and removed from the description. If it already exists the template will continue processing the other objects

File: templates/rhba70-full-mysql.yaml (hunk 5 above)
Comment (calvinzhuca) [reply templates/rhba70-full-mysql.yaml]: ok, updated based on you two's feedback

File: templates/rhba70-full-mysql-persistent.yaml [hunk 7]
```diff
@@ -280,7 +280,6 @@ objects:
       application: "${APPLICATION_NAME}"
     annotations:
       description: The Business Central web server's http port.
-      service.alpha.openshift.io/dependencies: '[{"name": "${APPLICATION_NAME}-mysql", "kind": "Service"}]'
```
Comment (errantepiphany) [inline templates/rhba70-full-mysql-persistent.yaml]: Curious why the removal of the dependencies (here and 3 other places)? Don't we want to make sure the mysql service is available before others that depend on it?

Comment (ruromero) [reply templates/rhba70-full-mysql-persistent.yaml]: @errantepiphany this was removed from the console which was its only purpose. See https://github.com/openshift/origin-web-console/issues/2319

Comment (calvinzhuca) [reply templates/rhba70-full-mysql-persistent.yaml]: ok, I squashed the 7 commits

Comment (errantepiphany) [reply templates/rhba70-full-mysql-persistent.yaml]: @ruromero Okay, thanks. I thought it was an actual service dependency (not just a console UI thing), but after consulting a couple other people, that doesn't seem to be the case.

**Pull request merged**